### PR TITLE
Make the interaction between #line and #nowarn directives consistent

### PR DESCRIPTION
**Pull request by Martin521**
## Description

This fix makes the interaction between `#line` and `#nowarn` directives consistent and defined.

Currently this is not the case, as the following example shows. 
Assume this file `xyz.fs`:

```
namespace X
#line 10 "xyz.fsy"
#nowarn "25"
#line 4 "xyz.fs"
module B = match None with None -> ()
```

It produces a warning FS0025 for line 5. If you replace the `10` in the first `#line` directive by `1`, no warning is shown. (Use the compiler to test this. Your editor, using the background compiler, tells again a different story. But we will tackle that later.)

The reason for the inconsistency is the [checkFile flag](https://github.com/dotnet/fsharp/blob/d37a8ae8aea915e1819b34c7fcd49749d11a3723/src/Compiler/Driver/CompilerDiagnostics.fs#L2241), introduced more than 10 years ago. Originally probably a hack in a debugging situation, it survived reviews since, and ended up being used in further modules. When this flag is set to `false` (which it is in most situations), `#nowarn` processing relies on comparing line numbers of one file with line numbers of another file. Accidentally, it sometimes works.

There are in principle two ways to fix this. (Note that the F# spec says nothing about the interaction of `#nowarn` and `#line`.)

The first is to consider `#line` directives as irrelevant for the scope of `#nowarn`. So, in the above example, "incomplete match" warnings would be disabled from line 3 to the end of the file. To implement this, however, we would need bookkeeping of the original line numbers *and* the line numbers adjusted by `#line` directives. Which would mean to extend the ubiquitous `range` struct to carry both kinds of line numbers. The cost of this (both refactoring cost and compiler runtime cost) is prohibitive.

The other other way is as follows. The sections that in terms of the `#line` directives belong to a certain file are considered by `#nowarn` as a separate file. For the typical case of source generation, this means that, in the generated file, a `#nowarn` in a line pointing to the generating file is effective only in other such lines, and not in lines that are not affected by any `#line` directive. So, in the above example, the `#nowarn 25` belongs to `xyz.fsy` and does therefore not suppress the warning for the last line, which belongs to `lineNoWarn.fs`.

This is a breaking change (in that warnings might appear that were suppressed in previous language versions) and must therefore be put under a feature flag. With this flag, the new specification is applied only for new language versions. But for previous language versions, we still want to fix this absurd bug. We do this by implementing a behavior that is equivalent to always having `checkFile = false` (which is currently the case anyway in 4 out of 5 cases that use the flag) *and* omitting the nonsensical comparison of line numbers from different files, i.e. apply any `#nowarn` to the whole file. This is a slightly looser check for "nowarn" than currently, which means that theoretically, a warning that was emitted by (say) SDK 8.0 might be suppressed when using language version 8.0 from the new SDK. This should be acceptable.

> Note to reviewers:
The core change is in `DiagnosticsLoggerFilteringByScopedPragmas(...)`  in `CompilerDiagnostics.fs`.
The change in signature of that constructor is reflected in small changes in `CompilerDiagnostics.fsi`, `fsc.fs`, `ParseAndCheckInputs.fs`, `IncrementalBuild.fs`, `TransparentCompiler.fs`.
The new consistent behavior needs adaptation of the three `.fsy` files and of `fsyaccdriver.fs`.
The feature flag needs changes in `FSComp.txt` and `LanguageFeatures.fs/fsi`.
Tests are covered by `Nowarn.fs`, which is included now in `FSharp.Compiler.ComponentTests.fsproj`.

## Checklist

- [X] Test cases added
- [X] Release notes entry updated

### CODE REVIEW

**Review by vzarytovskii — APPROVED** [review]

File: buildtools/fsyacc/fsyaccdriver.fs [hunk 1]
```diff
@@ -199,6 +199,8 @@ let writeSpecToFile (generatorState: GeneratorState) (spec: ParserSpec) (compile
               writer.WriteLineInterface "module %s" s;
       
       writer.WriteLine "#nowarn \"64\";; // turn off warnings that type variables used in production annotations are instantiated to concrete type";
+      writer.WriteLine "#nowarn \"1182\"  // the generated code often has unused variable 'parseState'"
```
Comment (nojaf) [inline buildtools/fsyacc/fsyaccdriver.fs:202]: This change should also be made in https://github.com/fsprojects/FsLexYacc/

Comment (Martin521) [reply buildtools/fsyacc/fsyaccdriver.fs:202]: Created https://github.com/fsprojects/FsLexYacc/pull/218

**Review by edgarfgp — APPROVED** [review]
Thanks Martin.

**Comment by Martin521** [comment]
Any volunteer for the second authorized review?

**Comment by vzarytovskii** [comment]
A bunch of folks are off/away currently, will review it in few days

File: src/Compiler/Driver/CompilerDiagnostics.fs [hunk 2]
```diff
@@ -2299,17 +2300,12 @@ type PhasedDiagnostic with
...
-// because we install a filtering error handler on a file-by-file basis for parsing and type-checking.
-// However this is indicative of a more systematic problem where source-line
-// sensitive operations (lexfilter and warning filtering) do not always
-// interact well with #line directives.
 type DiagnosticsLoggerFilteringByScopedPragmas
-    (checkFile, scopedPragmas, diagnosticOptions: FSharpDiagnosticOptions, diagnosticsLogger: DiagnosticsLogger) =
+    (langVersion: LanguageVersion, scopedPragmas, diagnosticOptions: FSharpDiagnosticOptions, diagnosticsLogger: DiagnosticsLogger) =
     inherit DiagnosticsLogger("DiagnosticsLoggerFilteringByScopedPragmas")
 
+    let bugFixed = langVersion.SupportsFeature LanguageFeature.CheckFileBugFixed
```
Comment (KevinRansom) [inline src/Compiler/Driver/CompilerDiagnostics.fs:2307]: The feature name should not be : CheckFileBugFixed, the name will be meaningless in 10 years.
Propose:
     ConsistentNowarnLineDirectiveTreatment or  ConsistentNowarnLineDirectiveInteraction

**Review by KevinRansom — CHANGES_REQUESTED** [review]
Thanks this looks good, but please rename the feature name.

File: src/Compiler/Driver/CompilerDiagnostics.fs (hunk 2 above)
Comment (Martin521) [reply src/Compiler/Driver/CompilerDiagnostics.fs]: done

**Pull request merged**